### ISSUE TITLE
Unable to push hardhat contracts

### ISSUE THREAD

**Issue by fleupold** [Tenderly/tenderly-cli#31]
Hi, I'm trying to debug failed transactions from a local ganache network with a dapp that has been built using hardhat.

After checking out [this repo](https://github.com/gnosis/gp-v2-contracts) & and adding a ganache network to the `hardhat.config.ts` (see below for details), I run 
1. `yarn hardhat deploy --network ganache`
2. `tenderly init`
3.  `tenderly push --networks ganache` 

And am getting the following error:

> Couldn't read Hardhat build files at: deployments

This is my ganache config:

```js
ganache: {
  ...sharedNetworkConfig,
  url: `http://localhost:8545`,
},
```

The same applies without specifying any networks (despite the contracts being deployed on Rinkeby, xDAI and Mainnet).

Would you be able to provide any pointers as to what configuration I might be missing?

Thanks

**Comment by Malcador** [comment]
Hey, thanks for reaching out.

I've ran the command with the --debug flag and it looks like we're having problems reading some contract source files. The error I get is `node_modules/solc_0.7/proxy/Proxy.sol` while the actual path is `node_modules/hardhat-deploy/solc_0.7/proxy/Proxy.sol`. 

It seems that all the source paths generated by hardhat deploy reference the files without the hardhat-deploy prefix, and that breaks our path resolution. I have no workaround at this time, but I will keep looking into this and let you know!

**Comment by deussatori** [comment]
Hey, I am unable to push as well ! We're using hardhat only and we get the same error: 
unable to upload contracts: unable to get provider contracts: failed listing build files: open deployments: The system cannot find the file specified.
Couldn't read Hardhat build files at: deployments

Is there a workaround ? Let me know please. It's important.

**Comment by BogdanHabic** [comment]
Hey @deussatori, are you using yarn link/workspaces by any chance?

**Comment by deussatori** [comment]
Hey, no.
I did try with a (1) contract as basic as possible a (2) hardhat config as basic as possible - still no luck.
I believe it's looking into the wrong folder. When I renamed "artifacts" to "deployments" I get this error:

"encountered unexcepted error: runtime error: invalid memory address or nil pointer dereference
Encountered an unexpected error. This can happen if you are running an older version of the Tenderly CLI."

So I believe this is it either.

**Comment by BogdanHabic** [comment]
Can you run `tenderly <command> --output json --debug`

If you want to keep the output private, you can go to [Tenderly](https://dashboard.tenderly.co), start the chat in the bottom right and paste it there 😃

**Comment by blakewest** [comment]
I also appear to be unable to deploy hardhat contracts. I'm using `hardhat-deploy`. My deployments are in the `deployments` folder. The output I get is...
```
Setting up your project...
Analyzing Hardhat configuration...
Pushing Smart Contracts for project: goldfinch/goldfinch-protocol
No migrated contracts detected in build directory: /Users/blakewest/code/goldfinch/goldfinch-protocol/deployments. This can happen when no contracts have been migrated yet.
Some of the project pushes were not successful. Please see the list above
```

I ran it with the debug flags, and put it into [this gist](https://gist.github.com/blakewest/c9a2fd222d78761d4ca04659b66766b0). Hopefully that helps! Would love to use this feature with my project.

**Comment by ericnordelo** [comment]
The same is happening to me 5 months after

**Comment by samlaf** [comment]
Also running into this issue...
```
> tenderly contracts push --networks 42 --debug                                               (py39) 
Setting up your project...
Trying OpenZeppelin config path: networks.js
couldn't read new OpenzeppelinConfig config file
Trying buidler config path: buidler.config.js
couldn't read new Buidler config file
Trying hardhat config path: hardhat.config.js
unable to fetch config
 Couldn't read Hardhat config file
Trying hardhat ts config path: hardhat.config.ts
Making request
Got response with body
Analyzing Hardhat configuration...
Trying Hardhat config path: /home/samlaf/dev/ethereum/hardhat/hardhat.config.js
Trying Hardhat config path: /home/samlaf/dev/ethereum/hardhat/hardhat.config.ts
Pushing Smart Contracts for project: hardhat-test
We have detected the following Smart Contracts:
• Greeter
•  (not deployed to any network, will be used as a library contract)
• Greeter (not deployed to any network, will be used as a library contract)
•  (not deployed to any network, will be used as a library contract)
[          ] Making request
[===       ] Got response with body
Push for hardhat-test failed with error: unexpected number of pushed contracts. Got: 0 expected: 4
Some of the contracts haven't been pushed. This can happen when the contract isn't deployed to a supported network or some other error might have occurred. Below is the list with all the contracts that weren't pushed successfully:
• Greeter on network 42 with address 0xB796bB43543A62d7d79C14aB6eA36A65716bc74B
unable to upload contracts: some project uploads failed
Some of the project pushes were not successful. Please see the list above
```
Just trying to push the simple Greeter contract that comes from initializing a fresh hardhat typescript project.

**Comment by samlaf** [comment]
Thanks @BogdanHabic, I moved on from this issue, but will make sure to do contact you next time I need to debug a transaction using tenderly. :)

**Comment by Donaldson7** [comment]
Yeah really can't resimulate transactions directly to the right address and
ain't enough funds to pay my gas prices

**Comment by BogdanHabic** [comment]
Hey @Donaldson7, what do you mean that you can't resimulate transactions to the right address? All addresses are unlocked by default in the Simulator and Forks.

**Comment by spaceh3ad** [comment]
I also have error about not-existant deploymets folder. I don't understand I don't have such folder in my hardhat project why tenederly is asking for one? Neither in  truffle .

**Comment by tantely-npx** [comment]
As anyone managed to solve this issue ? I'm trying to push a contract from Hardhat but facing this issue:
`Couldn't read Hardhat build files at: deployments`

**Comment by Malcador** [comment]
Hey, have you checked out [this](https://github.com/Tenderly/tenderly-cli/issues/71#issuecomment-1022348594)? It should help with the issue you guys are facing.

Tenderly works with artifacts generated by either `hadthat-deploy` or `hardhat-tenderly` plugins, so you need one of those in order for the required directories to be generated.

**Comment by tantely-npx** [comment]
> 

Thanks it helped me solve the issue 👍

**Comment by sparrowDom** [comment]
Hey @BogdanHabic I am also facing this error when I try to push a local transaction off of my local node (running in forked mode). It finds the hardhat config, and I believe that after it tried to read contract artifacts it throws the: 

`encountered unexcepted error: runtime error: invalid memory address or nil pointer dereference`

Happy to debug with you if you are willing

**Comment by beruda** [comment]
@sparrowDom We're moving away from supporting local transactions, and moving towards our own local development solution: DevNets (read more [here](https://docs.tenderly.co/devnets/intro-to-devnets)).

This is the gist (generated by GitBook Lens):

> DevNets are managed, zero-setup environments for smart contract development. They provide an isolated and flexible network where you can deploy, test, and debug your smart contracts using tools like the Debugger, Gas Profiler, and Call Trace. DevNets also allow you to access Mainnet data, simulate transactions, and customize the blockchain environment. They offer benefits such as zero-setup, isolation, flexibility, unlimited faucet, and reusable templates.

I hope this helps!
